openverse_catalog/dags/providers/provider_api_scripts/nypl.py

```diff
@@ -1,210 +1,229 @@
-
-    return image_url
-
-
-def _get_preferred_image(image_type, dimension_list):
-    preferred_image = (
-        image_type.get(dimension).replace("&download=1", "")
-        for dimension in dimension_list
-        if dimension in image_type
-    )
-
-    return next(preferred_image, None)
-
-
-def _get_metadata(mods):
-    metadata = {}
-
-    type_of_resource = mods.get("typeOfResource")
-    if isinstance(type_of_resource, list) and (
-        type_of_resource[0].get("usage") == "primary"
-    ):
-        metadata["type_of_resource"] = type_of_resource[0].get("$")
-
-    if isinstance(mods.get("genre"), dict):
-        metadata["genre"] = mods.get("genre").get("$")
-
-    origin_info = mods.get("originInfo")
-    if date_issued := origin_info.get("dateIssued", {}).get("$"):
-        metadata["date_issued"] = date_issued
-    if publisher := origin_info.get("publisher", {}).get("$"):
-        metadata["publisher"] = publisher
+        image_url, filetype = next(preferred_image, None)
+        image_url = image_url.replace("http://", "https://")
+        return image_url, filetype
+
+    @staticmethod
+    def _get_creators(creatorinfo):
+        if isinstance(creatorinfo, list):
+            primary_creator = (
+                info.get("namePart", {}).get("$")
+                for info in creatorinfo
+                if info.get("usage") == "primary"
+            )
+            creator = next(primary_creator, None)
+        else:
+            creator = creatorinfo.get("namePart", {}).get("$")
+
+        return creator
+
+    @staticmethod
+    def _get_metadata(mods):
+        metadata = {}
+
+        type_of_resource = mods.get("typeOfResource")
+        if isinstance(type_of_resource, list) and (
+            type_of_resource[0].get("usage") == "primary"
+        ):
+            metadata["type_of_resource"] = type_of_resource[0].get("$")
+
+        if isinstance(mods.get("genre"), dict):
+            metadata["genre"] = mods.get("genre").get("$")
+
+        origin_info = mods.get("originInfo", {})
+        if date_issued := origin_info.get("dateIssued", {}).get("$"):
+            metadata["date_issued"] = date_issued
+        if date_created := origin_info.get("dateCreated", {}):
+            if isinstance(date_created, list):
+                # Approximate dates have a start and an end
+                # [{'encoding': 'w3cdtf', 'keyDate': 'yes',
+                # 'point': 'start', 'qualifier': 'approximate', '$': '1990'},
+                # {'encoding': 'w3cdtf', 'point': 'end',
+                # 'qualifier': 'approximate', '$': '1999'}]
+                start_date = (
+                    d.get("$") for d in date_created if d.get("point") == "start"
+                )
+                end_date = (d.get("$") for d in date_created if d.get("point") == "end")
+                if next(start_date) and next(end_date):
+                    date_created = f"{start_date}-{end_date}"
```

Comment:
I'm not sure this will work - `start_date` and `end_date` are generators so this will just interpolate the generator representation into `date_created` 😮 Do we have a test for this particular code path?